specifyweb/frontend/js_src/lib/components/WorkBench/WbView.tsx

```diff
@@ -142,6 +152,58 @@ export function WbView({
     spreadsheetContainerRef,
   });
 
+  const [disambiguationMatches, setDisambiguationMatches] = React.useState<{
+    readonly physicalCols: RA<number>;
+    readonly mappingPath: MappingPath;
+    readonly ids: RA<number>;
+    readonly key: string;
+  }>();
+  const [disambiguationPhysicalRow, setPhysicalRow] = React.useState<number>();
+  const [disambiguationResource, setResource] =
+    React.useState<Collection<AnySchema>>();
+  const [
+    noDisambiguationDialog,
+    openNoDisambiguationDialog,
+    closeNoDisambiguationDialog,
+  ] = useBooleanState();
+  const [disambiguationDialog, openDisambiguation, closeDisambiguation] =
+    useBooleanState();
+
+  const openDisambiguationDialog = React.useCallback(() => {
+    if (mappings === undefined || hot === undefined) return;
+
+    const [visualRow, visualCol] = getSelectedLast(hot);
+    const physicalRow = hot.toPhysicalRow(visualRow);
+    const physicalCol = hot.toPhysicalColumn(visualCol);
+
+    const matches = workbench.validation.uploadResults.ambiguousMatches[
+      physicalRow
+    ].find(({ physicalCols }) => physicalCols.includes(physicalCol));
+    if (matches === undefined) return;
+    const tableName = getTableFromMappingPath(
+      mappings.baseTable.name,
+      matches.mappingPath
+    );
+    const table = strictGetTable(tableName);
+    const resources = new table.LazyCollection({
+      filters: { id__in: matches.ids.join(',') },
+    }) as Collection<AnySchema>;
+
+    (hasTablePermission(table.name, 'read')
+      ? resources.fetch({ limit: 0 })
+      : Promise.resolve(resources)
+    ).then(({ models }) => {
+      if (models.length === 0) {
+        openNoDisambiguationDialog();
+        return;
+      }
+      setDisambiguationMatches(matches);
+      setResource(resources);
+      setPhysicalRow(physicalRow);
+      openDisambiguation();
+    });
+  }, [mappings, hot]);
```

Comment:
I would suggest you move it into a separate hook in a separate file

that hook would receive these arguments: mappings, hot, 

then, all of these would also be moved into the hook:

```ts
  const [disambiguationMatches, setDisambiguationMatches] = React.useState<{
    readonly physicalCols: RA<number>;
    readonly mappingPath: MappingPath;
    readonly ids: RA<number>;
    readonly key: string;
  }>();
  const [disambiguationPhysicalRow, setPhysicalRow] = React.useState<number>();
  const [disambiguationResource, setResource] =
    React.useState<Collection<AnySchema>>();
  const [
    noDisambiguationDialog,
    openNoDisambiguationDialog,
    closeNoDisambiguationDialog,
  ] = useBooleanState();
  const [disambiguationDialog, openDisambiguation, closeDisambiguation] =
    useBooleanState();
``

and if some of these are needed outside the hook, the hook can just return an object that contains the values that are needed

but rather than returning the states directly, the hook could return JSX too, for example, this code could be moved into a hook:
```tsx
   {noDisambiguationDialog && (
        <Dialog
          buttons={commonText.close()}
          header={wbText.noDisambiguationResults()}
          onClose={closeNoDisambiguationDialog}
        >
          {wbText.noDisambiguationResultsDescription()}
        </Dialog>
      )}
      {disambiguationDialog && (
        <DisambiguationDialog
          matches={disambiguationResource!.models}
          liveValidationStack={workbench.validation.liveValidationStack}
          onClose={closeDisambiguation}
          onSelected={(selected) => {
```